onnxruntime/contrib_ops/cpu/bert/gqa_attention_base.h

```diff
@@ -242,35 +268,49 @@ class GQAAttentionBase {
+
+          const bool should_apply_local_window = local_window_size_ > 0 &&
+                                                 seq_causal_length > static_cast<size_t>(local_window_size_) + 1;
+
+          const size_t start_offset = should_apply_local_window ? seq_causal_length - local_window_size_ - 1 : 0;
+          const size_t window_size = should_apply_local_window ? local_window_size_ + 1 : seq_causal_length;
+
+          // Mask everything before local window, if local window should be applied
+          if (should_apply_local_window) {
             for (size_t total_seq_id = 0; total_seq_id < seq_causal_length - local_window_size_ - 1; total_seq_id++) {
               if constexpr (std::is_same<U, float>::value) {
                 output_softmax[total_seq_id] = 0.f;
               } else {
                 output_softmax[total_seq_id] = MLFloat16::FromBits(static_cast<uint16_t>(0));
               }
             }
-            if (softcap_ > 0.f) {
-              ComputeAttentionSoftcapInplace(output_softmax + seq_causal_length - local_window_size_ - 1,
-                                             local_window_size_ + 1, static_cast<U>(softcap_));
-            }
-            if (use_smooth_softmax_) {
-              ComputeSmoothSoftmaxInplace(output_softmax + seq_causal_length - local_window_size_ - 1, 1,
-                                          local_window_size_ + 1, nullptr);
+          }
+
+          if (softcap_ > 0.f) {
+            ComputeAttentionSoftcapInplace(output_softmax + start_offset, static_cast<int>(window_size),
+                                           static_cast<U>(softcap_));
+          }
+
+          // Apply custom attention mask if there is any
+          // TODO (#23982): Implement masking during softmax computation in GQA CPU operator
+          if (attention_mask_thread != nullptr) {
+            if constexpr (std::is_same_v<U, T>) {
+              ApplyAttentionMask(output_softmax + start_offset, attention_mask_thread + start_offset,
+                                 static_cast<int>(window_size));
             } else {
-              ComputeAttentionSoftmaxInplace(output_softmax + seq_causal_length - local_window_size_ - 1, 1,
-                                             local_window_size_ + 1, nullptr);
+              size_t bytes = window_size * sizeof(float);
+              auto attention_mask_thread_fp32 = static_cast<float*>(allocator->Alloc(bytes));
+              BufferUniquePtr scratch_buffer(attention_mask_thread_fp32, BufferDeleter(allocator));
+
+              MlasConvertHalfToFloatBuffer(attention_mask_thread + start_offset, attention_mask_thread_fp32, window_size);
+              ApplyAttentionMask(output_softmax, attention_mask_thread_fp32, static_cast<int>(window_size));
             }
+          }
+
+          if (use_smooth_softmax_) {
+            ComputeSmoothSoftmaxInplace(output_softmax + start_offset, 1, static_cast<int>(window_size), nullptr);
```

Comment:
Please move the logic right after attention masking:
if (use_smooth_softmax_) {
    ComputeSmoothSoftmaxInplace
} else {
    ComputeAttentionSoftmaxInplace
}
We shall not do a sequence of softmax + smooth_softmax.